symengine/expansion.cpp

```diff
@@ -0,0 +1,619 @@
+        throw std::logic_error("atanh(const) not Implemented");
+
+    const std::string &varname = var->get_name();
+    const pp_t& p(-(s.pow(2)) + 1);
+    pp_t::set_auto_truncate_degree(prec-1, {varname});
+    pp_t res_p = s.partial(varname) * series_invert(p, var, prec-1);
+    pp_t::set_auto_truncate_degree(prec, {varname});
+    return res_p.integrate(varname);
+    return res_p;
+}
+
+pp_t series_asinh(const pp_t s, const RCP<const Symbol> &var, unsigned int prec)
+{
+    if (s.find_cf({0}) != 0)
+        throw std::logic_error("asinh(const) not Implemented");
+
+    const std::string &varname = var->get_name();
+    const pp_t& p(series_nthroot(s.pow(2) + 1, 2, var, prec));
+    pp_t::set_auto_truncate_degree(prec-1, {varname});
+    pp_t res_p = s.partial(varname) * series_invert(p, var, prec-1);
+    pp_t::set_auto_truncate_degree(prec, {varname});
+    return res_p.integrate(varname);
+    return res_p;
+}
+
+pp_t series_tanh(const pp_t s, const RCP<const Symbol> &var, unsigned int prec)
+{
+    if (s.find_cf({0}) != 0)
+        throw std::logic_error("tanh(const) not Implemented");
+
+    const std::string &varname = var->get_name();
+    pp_t res_p{s};
+    auto steps = step_list(prec);
+    for (const auto step : steps) {
+        pp_t::set_auto_truncate_degree(step, {varname});
+        const pp_t p(s - series_atanh(res_p, var, step));
+        res_p += -p * (res_p.pow(2) - 1);
+    }
+    return res_p;
+}
+
+using series_func_t = decltype(&series_tan);
+
+static std::map<TypeID, series_func_t> funcmap;
+void build_funcmap() {
+    RCP<const Symbol> x = symbol("x");
+    funcmap[tan(x)->get_type_code()] = &series_tan;
+    funcmap[cot(x)->get_type_code()] = &series_cot;
+    funcmap[csc(x)->get_type_code()] = &series_csc;
+    funcmap[sec(x)->get_type_code()] = &series_sec;
+    funcmap[sin(x)->get_type_code()] = &series_sin;
+    funcmap[cos(x)->get_type_code()] = &series_cos;
+    funcmap[log(x)->get_type_code()] = &series_log;
+    funcmap[atan(x)->get_type_code()] = &series_atan;
+    funcmap[asin(x)->get_type_code()] = &series_asin;
+    funcmap[sinh(x)->get_type_code()] = &series_sinh;
+    funcmap[cosh(x)->get_type_code()] = &series_cosh;
+    funcmap[atanh(x)->get_type_code()] = &series_atanh;
+    funcmap[asinh(x)->get_type_code()] = &series_asinh;
+    funcmap[tanh(x)->get_type_code()] = &series_tanh;
+    funcmap[lambertw(x)->get_type_code()] = &series_lambertw;
+}
+
+pp_t num2poly(const Number &num)
+{
+    if (is_a<Integer>(num)) {
+        const Integer& ii = static_cast<const Integer&>(num);
+        const piranha::integer bigint(ii.i.get_mpz_t());
+        return pp_t(bigint);
+    }
+    throw std::runtime_error("Unhandled type in num2poly");
+}
+
+//=================================================================
+//! Transform the Basic expression and return the series polynomial
+pp_t _series(const RCP<const Basic> &ex, const RCP<const Symbol> &var, unsigned int prec)
+{
+    if (is_a<Integer>(*ex)) {
```

Comment:
Checking using `is_a` was slow for most other applications, but it depends on which operation takes the most time.
`eval_double` and `eval_double_single_dispatch` are examples of two ways that tries to be faster than type checking.